### ISSUE TITLE
[Troubleshooting] Redirect from okta hits a 400, but has a valid URI

### ISSUE THREAD

**Issue by jskrzypek** [vouch/vouch-proxy#424]
I just spent an hour writing a thorough issue and github deleted it :cursing_face:. So this time I'm going to be a bit more brief, sorry. I can elaborate as needed.

### Observed behavior
I do the vouch redirect flow through okta and back and get redirected to a Vouch Proxy 400 error page. On this error page I click on the URL box and *without changing anything* hit enter, and then the vouch proxy flow completes and I see the intended page that is served behind vouch.

Here's a screenshot of the 400 error page: 
![image](https://user-images.githubusercontent.com/1513265/133141726-216d66c2-169a-4f9d-98f4-b9f0aa35766f.png)

### Expected behavior
Same as observed, but I should not have to stop on the 400 page in the middle.

### Setup
Vouch & the app it wraps are served with k8s, and vouch proxy comes from the hawkeye helm chart (with an updated Ingress to v1 instead of the beta -- we're on a newer k8s api version). I can post the sanitized configuration if requested, but I don't think it's the issue...

I'm using the `alpine-0.34.1` docker image tag from quay.io, and using the new vouch-in-path feature with `vouch.domain_root: /sso`. I was seeing this issue before but wasn't sure if it was related so I was waiting to see if that resolved the issue (it didn't). `vouch.testing` is set to `false` for obvious reasons.

### Context
- My laptop is connected to our VPN and accessing the app & vouch-proxy from there. 
- OS: Linux (NixOS) & MacOS
- Browsers used:
  - Chromium - 92.0.4515.159 (Official Build) (64-bit)
    - Need to completely log out of the whole browser/okta session to reproduce after the first time. Testing in incognito mode was easier. The logs below were generated when I did the flow from chromium. 
  - Firefox - 91.0.1 (64-bit)
    - Was able to reproduce several times without blowing away the entire browser session, just by deleting the cookies that were present.

### Sanitized & Annotated Logs
I can reproduce and repost the logs if needed. I have annotated what I believe was happening during the described flow based on the timestamps. I am uncertain if `*** GAP 1 ***` or `*** GAP 2 ***` corresponds to the time it took me to click on the url box and hit enter. I think this would only take about 800ms rather than 2.3s, but it's hard to say (I was sort of practiced after having done it several times already), and either could make sense.
https://hastebin.com/raw/rilewecoma

**Comment by bnfinet** [comment]
@jskrzypek Could you please provide your config as described in the README, turn on `testing` and point to any documentation for your ingress setup.  Please also provide full logs as per the README, not just pieces here and there.

I think that you're not being redirected back to your original URL properly.  That could be an Nginx configuration or something else.

What's the URL you try to access directly to being the flow?

This log is interesting since it's the VP logo in the 400 document.  Is `/sso/static` setup as a route?
`session requestedURL set to https://[[SANITIZED APP DOMAIN]]/static/img/multicolor_V_500x500.png`

I'm not a current k8s user so I'm not up to date and nginx Ingress.

FWIW your logs show VP working properly. `/auth/$STATE` makes the proper UserInfo call and returns a cookie.  After that `/validate` evaluates the cookie and responds properly.

**Comment by jskrzypek** [comment]
> Could you please provide your config as described in the README, turn on testing and point to any documentation for your ingress setup.

Yeah, sure I'll capture and post those soon. Doesn't `vouch.testing` cause vouch to not actually redirect the browser, just print out the 302 messages?

>  Please also provide full logs as per the README, not just pieces here and there.
I didn't cut anything out of the snippet, just inserted the lines with the notes as to what I was doing when. the rest of the logs are just a bunch of `/healthcheck` calls that are part of the normal k8s operation (i.e. k8s is checking that the pod hosting the server is still healthy). ***Much*** earlier you can see the boot-up logs from Vouch. Are those the ones you want?

> I think that you're not being redirected back to your original URL properly. That could be an Nginx configuration or something else. 

Sure that makes sense, I was just not sure how to dig further into the options, so I posted

> What's the URL you try to access directly to being the flow?
The same thing happens if I first navigate to either of `https://<domain-hidden>` or `https://<domain-hidden>/index` (or anything else really).

> Is /sso/static setup as a route?

Yes, I can access `https://<domain-hidden>/sso/static/img/multicolor_V_500x500.png`, even when not authenticated.

OTOH, when I inspect that broken image on the page it's expecting it to live at `https://<domain-hidden>/static/img/multicolor_V_500x500.png`. This may be an issue, but I think it's not since at the time that it can't find that image, it's already trying to render the 400 error page.

> FWIW your logs show VP working properly. 
Yes, VP works properly b/c the exact same URL that resolves as a 400 error page on the initial redirect works after I manually trigger a page load

**Comment by jskrzypek** [comment]
Also these logs specifically seem to indicate where the 400 error is happening (this is from that same logfile I posted above).
```
{"level":"debug","ts":1631555404.4817154,"msg":"/auth"}
{"level":"debug","ts":1631555404.4817688,"msg":"CaptureWriter.Write set w.StatusCode 302"}
{"level":"info","ts":1631555404.481917,"msg":"|302|  110.951µs /sso/auth","statusCode":302,"request":1206,"latency":0.000110951,"avgLatency":0.007823254,"ipPort":"[[SANITIZED CLIENT IP 1]]:49186","method":"GET","host":"[[SANITIZED APP DOMAIN]]","path":"/sso/auth","referer":""}
{"level":"debug","ts":1631555404.5206268,"msg":"/auth/{state}/"}
{"level":"warn","ts":1631555404.5207055,"msg":"/auth Invalid session state: stored %!s(<nil>), returned [[SANITIZED VOUCH SESSION STATE 1]]"}
{"level":"debug","ts":1631555404.5207143,"msg":"setting the cookie domain to [[SANITIZED APP DOMAIN]]"}
{"level":"debug","ts":1631555404.5207298,"msg":"rendering error for user: 400 Bad Request"}
{"level":"debug","ts":1631555404.5207357,"msg":"CaptureWriter.Write set w.StatusCode 400"}
{"level":"info","ts":1631555404.5209327,"msg":"|400|  192.228µs /sso/auth/[[SANITIZED VOUCH SESSION STATE 1]]/","statusCode":400,"request":1207,"latency":0.000192228,"avgLatency":0.007816932,"ipPort":"[[SANITIZED CLIENT IP 1]]:49186","method":"GET","host":"[[SANITIZED APP DOMAIN]]","path":"/sso/auth/[[SANITIZED VOUCH SESSION STATE 1]]/","referer":""}
```

**Comment by jskrzypek** [comment]
When I enable `VOUCH_TESTING=true`, I am able to click through the 302 redirect pages without errors (the logo also loads correctly but I suspect that's not related), and that message I pasted above doesn't appear (i.e. the `/auth Invalid session state: stored %!s(<nil>), returned [[SANITIZED VOUCH SESSION STATE 1]]`).

When I disable it again with `VOUCH_TESTING=false` that `/auth Invalid session state` message shows up again.

**Comment by bnfinet** [comment]
@jskrzypek without seeing your configuration and complete logs (as documented in the README) it's nearly impossible to advise you properly.

**Comment by jskrzypek** [comment]
@bnfinet sorry maybe I should have called it out separately. 

I asked before: what is more useful for you to see? The kubernetes yaml configs, or the resultant nginx.conf file that gets produced by those configs?

**Comment by bnfinet** [comment]
@jskrzypek my apologies I missed that ask.  Is it difficult to offer both?
